search/src/Filtering.cpp

```diff
@@ -46,5 +47,207 @@ std::vector<int> sigmaGFilteredIndices(const std::vector<float>& values,
     }
     return result;
 }
+
+/* Given a set of psi and phi values,
+   return a likelihood value */
+double calculateLikelihoodFromPsiPhi(std::vector<double> psiValues, std::vector<double> phiValues)
+{
+    double psiSum = 0.0;
+    double phiSum = 0.0;
+    
+    for(int i = 0; i < psiValues.size(); i++) {
+        psiSum += psiValues[i];
+        phiSum += phiValues[i];
+        
+    }
+    
+    if(psiSum == 0.0 || phiSum <= 0.0) {
+        return 0.0;
+    }
+    
+    return psiSum / sqrt(phiSum);
+}
+    
+/* Given a set of fluxValues,
+   inverse Phi values, and a
+   vector of positive flux indicies,
+   apply the kalman filters to 
+   said positive values. */
+std::tuple<std::vector<double>, std::vector<double>> calculateKalmanFlux(std::vector<double> fluxValues, 
+                                                                         std::vector<double> vars,
+                                                                         std::vector<int> fluxIdx, bool reverse)
+{
+    int fluxSize = fluxValues.size();
+    double kalmanFluxes[fluxSize];
+    
+    std::vector<double> xhat(fluxSize, 0.0);
+    std::vector<double> p(fluxSize, 0.0);
+    double xhatMinus[fluxSize];
+    double pMinus[fluxSize];
+    double k[fluxSize];
+    
+    double q = 1.0;
+    
+    /* If reverse is false, calculate
+       Kalman result going 'forward'
+       (0 -> end of fluxes)
+       Else, calculate it 'backwards'
+       (end of fluxes -> 0)*/
+    if(!reverse) {
+        xhat[fluxIdx[0]] = fluxValues[fluxIdx[0]];
+        p[fluxIdx[0]] = vars[fluxIdx[0]];
+        
+        for(int i = 1; i < fluxIdx.size(); i++) {
+            int idx = fluxIdx[i];
+            int idxMinus = fluxIdx[i - 1];
+            
+            xhatMinus[idx] = xhat[idxMinus];
+            pMinus[idx] = p[idxMinus] + q;
+
+            k[idx] = pMinus[idx] / (pMinus[idx] + vars[idx]);
+            xhat[idx] = xhatMinus[idx] + k[idx]*(fluxValues[idx]-xhatMinus[idx]);
+            p[idx] = (1.0-k[idx])*pMinus[idx];
```

Comment:
Are lines 107-109 the same as 121-123? And if so, would it be beneficial to turn them into a separate function?

Reply:
in the original python code, the fluxes are passed through the kalman filter once and then all the flux and variance values are reversed and passed through again. In the c implementation I decided against doing that to save on time/memory complexity. I originally tried doing this by writing the loop once and having the 'direction of calculation' handled by an iteration variable but it ended up being really hard to read, so for code clarity's sake I just split it into two similar loops, forwards and backwards. definitely open to changing up the structure if y'all have a better idea since I agree that it's kind of ugly 😭